src/nodesessionmetadata.cpp

```diff
@@ -106,4 +116,31 @@ session_id_t NodeSessionMetadata::getSubsessionSize(const std::string& subsessio
     }
     return std::get<1>(it->second);
 }
+
+session_id_t NodeSessionMetadata::getShardId(const std::set<std::string>& collapsedNames) const {
+    if (collapsedNames.size() == 0) {
+        return 1;
+    }
+    SPDLOG_ERROR("sessionsLebels.size(): {}, size-1: {}", sessionsLevels.size(), sessionsLevels.size() - 1);
+    if (collapsedNames.size() > sessionsLevels.size()) {
+        SPDLOG_LOGGER_ERROR(dag_executor_logger, "Tried to collapse more subsession levels than exists");
+        throw std::logic_error("Tried to collapse more subsession levels than exists");
+    }
+    for (size_t i = sessionsLevels.size() - 1; i > 0; --i) {
+        if (collapsedNames.find(sessionsLevels[i]) == collapsedNames.end()) {
+            SPDLOG_LOGGER_ERROR(dag_executor_logger, "Tried to collapse sessions not in LIFO order. Should collapse: {} first", sessionsLevels[i]);
+            throw std::logic_error("Cannot collapse sessions not in LIFO order");
+        }
+    }
+    session_id_t multiplyFactor = 1;
+    session_id_t shardId = 0;
+    for (size_t i = 0; i < collapsedNames.size(); ++i) {
+        const auto subsessionDetails = details.at(*(sessionsLevels.rbegin() + i));
+        const auto [id, sessionSize] = subsessionDetails;
```

Comment:
Do we need to copy here aswell? Wouldn't `const auto& [id, sessionSize]` extract from pair by reference?